File: kangaroo/kangaroo.go
package kangaroo

func kangaroo(aPos int32, aJumpSize int32, bPos int32, bJumpSize int32) string {
	var behindJumpSize int32
	var aheadJumpSize int32
	if aPos <= bPos {
		behindJumpSize = aJumpSize
		aheadJumpSize = bJumpSize
	} else {
		behindJumpSize = bJumpSize
		aheadJumpSize = aJumpSize
	}

	if aheadJumpSize >= behindJumpSize {
		return "NO"
	}

	var behindPos int32
	var aheadPos int32
	if aPos <= bPos {
		behindPos = aPos
		aheadPos = bPos
	} else {
		behindPos = bPos
		aheadPos = aPos
	}

	isWinnable := winnable(
		behindPos,
		behindJumpSize,
		aheadPos,
		aheadJumpSize,
	)

	if isWinnable {
		return "YES"
	} else {
		return "NO"
	}
}

func winnable(behindPos int32, behindJumpSize int32, aheadPos int32, aheadJumpSize int32) bool {
	if aheadPos <= behindPos {
		return aheadPos == behindPos
	} else {
		return winnable(
			behindPos+behindJumpSize,
			behindJumpSize,
			aheadPos+aheadJumpSize,
			aheadJumpSize,
		)
	}
}

File: kangaroo/Kangaroo.java
package kangaroo;

import java.util.Map;

public class Kangaroo {

  public static String kangaroo(int aPos, int aJumpSize, int bPos, int bJumpSize) {
    Map<Integer, Integer> jumpSizeByPosition = Map.of(
      aPos, aJumpSize,
      bPos, bJumpSize
    );

    int behindJumpSize = aPos <= bPos
      ? jumpSizeByPosition.get(aPos)
      : jumpSizeByPosition.get(bPos);

    int aheadJumpSize = aPos <= bPos
      ? jumpSizeByPosition.get(bPos)
      : jumpSizeByPosition.get(aPos);

    if (aheadJumpSize >= behindJumpSize) {
      return "NO";
    }

    int behindPos = Math.min(aPos, bPos);
    int aheadPos = Math.max(aPos, bPos);

    return isWinnable(behindPos, behindJumpSize, aheadPos, aheadJumpSize)
        ? "YES"
        : "NO";
  }

  private static boolean isWinnable(int behindPos, int behindJumpSize, int aheadPos, int aheadJumpSize) {
    return aheadPos <= behindPos
        ? aheadPos == behindPos
        : isWinnable(
            behindPos + behindJumpSize,
            behindJumpSize,
            aheadPos + aheadJumpSize,
            aheadJumpSize);
  }

}
